Use tidy eval with colpair_map()

jameslairdsmith opened the pull request:
#127 pointed out that `colpair_map()` behaves incorrectly when one of the columns has names that are the same as the those of the function's (underlying) arguments. The example given was with a column called `x`. The fix since made in #129 is helpful: it makes a problem much less likely by renaming that argument `x` to `.data_col` and the `data` argument of the helper function to `.data`.

But the function still fails if a column is named `f`:

``` r
# devtools::install_github("tidymodels/corrr") 
library(dplyr, warn.conflicts = F)
library(corrr)

mtcars %>% 
  rename(f = mpg) %>% 
  colpair_map(cor)
#> Error: Problem with `summarise()` input `..1`.
#> x Problem with `across()` input `.fns`.
#> ℹ Input `.fns` must be NULL, a function, a formula, or a list of functions/formulas.
#> ℹ Input `..1` is `dplyr::across(...)`.
```

<sup>Created on 2021-01-30 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

This PR makes a robust fix by tunnelling the various argument names using the double braces `{{}}` as necessary. This means the tidy evaluation works appropriately and we no longer have to worry about what users name their columns. I've named `.data_col` back to `x` and `.data` back to `data`, as they were before.

To confirm the fix is robust, I've also added a test (also shown below) where I rename the columns of mtcars to all the argument names used in the function and check whether we still get the same result we would get with `correlate()`:

``` r
library(waldo)
library(corrr)

arg_names <- c("x", ".x", "f", ".f" , ".fns", "data", ".data",
               "summarise_col", ".cols", "diagonal", ".diagonal")

my_mtcars <- mtcars
colnames(my_mtcars) <- arg_names

compare(colpair_map(my_mtcars, cor),
        correlate(my_mtcars, quiet = TRUE),
        tolerance = 1e-10)
#> ✓ No differences
```

<sup>Created on 2021-01-30 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

juliasilge commented:
This is currently failing because of the change introduced in dplyr 1.0.4 to `across()` noted in tidyverse/dplyr#5734 that is now fixed. The new version of dplyr will be submitted to CRAN in the next day or so; when that goes through, let's recheck and then merge in!

Thank you so much @jameslairdsmith  and @thisisdaryn! 🙌

thisisdaryn commented:
I discussed this PR with @juliasilge and we decided to keep the same variable names across the functions for maintainability/readability.

jameslairdsmith commented:
Thanks @juliasilge and @thisisdaryn! We will wait for that change to dplyr go through. I also spotted and fixed one tiny punctuation mistake which I've tagged on to this.

juliasilge commented:
The new version of [dplyr just made it to CRAN](https://cran.r-project.org/package=dplyr). 🎉 

I tried re-running the CI but the new version has of course not been built and propagated out yet; let's try on Monday.

jameslairdsmith commented:
> The new version of [dplyr just made it to CRAN](https://cran.r-project.org/package=dplyr). 🎉
> 
> I tried re-running the CI but the new version has of course not been built and propagated out yet; let's try on Monday.

Thanks @juliasilge! It may be that some it takes some time to propagate, but I had a look locally after installing dplyr 1.0.4 and there was one more issue which was causing the test to fail.

I've made two further commits and I explain more on them in the commit messages:

- One thing that happened with the new version of dplyr is that the tidy eval changes I made became redundant. You could run the test without them and it would still pass. This is good obviously, because it shows it's safer. I've kept the tidy eval fixes there for compatibility with those earlier versions of dplyr. And I've added comments to say why they are there.
- One thing that broke in the new version of dplyr is that there is no way (that I can find) to allow for `.data` to be a column name. As I explain in the commit message, I think this is fine. It's hugely unlikely to be a problem for any user. So I've removed `.data` from the set of legitimate column names we test can work.

thisisdaryn commented:
Merged. Thanks @jameslairdsmith!
